### ISSUE TITLE
Write the "presence protocol for opportunistic syncing" in JavaScript.

### ISSUE THREAD

**Issue by srichal** [thaliproject/thali#161]
Here is the spec that describes the protocol:
https://github.com/thaliproject/thali/blob/yaronyg-patch-1/presenceprotocolforopportunisticsynching.md

Here is the Java code that implements the protocol:
https://github.com/yaronyg/cryptopresencetest

This task is for implementing the "Generating the pre-amble and beacons" part of the protocol in JS. If any of the needed crypto APIs are not available in Node.js, a 3rd party library will be used.

**Comment by srichal** [comment]
The goal:
Write "generateDiscoveryAnnouncement()" and all the corresponding functions from the "DiscoveryAnnouncement" java-class in JavaScript.

The Test:
For the same set of "kxListOfReceivingDevicesPublicKeys", the java & JS code should return the same values.

**Comment by srichal** [comment]
Status as of Friday-15th:
1) Completed writing all the functions needed to generate the discovery announcement in JS.
2) Tested the below functions in JS:
          createKeyPair()
          createExpiration()
          generateUnencryptedKeyId()

Goal for Friday 22nd:
1) Test the below JS functions:
          generateSalt()
          aesAndKeyId()
          generateInsideBeaconHmac()
2) Add generateECDHWithHKDF() function in C.
3) Test the Java & JS functions that generate the discovery announcement using the same input parameters.

**Comment by srichal** [comment]
Status as of Friday-22nd:
1) Completed testing all the functions needed to generate the discovery announcement in JS.

Goal for Friday 29th: (working on tue/wed/thu only)
Write and test the C-API "generateECDHWithHKDF()" as this functionality is not available in Nodejs/crypto.

**Comment by srichal** [comment]
Status as of Friday-29th:
1) Wrote and tested the C-API "generateECDHWithHKDF()" as this functionality is not available in Nodejs/crypto.
2) Also tested calling this C-API from JS code. Was able to pass all the required the values to the C side and was able to get the generated value back to the JS side.

Goal for Wednesday-4th: (working on mon/tue/wed only)
JXcore does not support all the crypto functions that we need to call in order to generate the pre-amble and beacons. Need to find NPM modules as substitutes.

**Comment by srichal** [comment]
Status as of Wednesday-3rd:
Started integrating all my crypto work in C as well as JS to complete the Thali requirements under "0.0 - Bring in the public keys!". So far completed:
1) Generate a public/private key pair and return a PKCS#12 encoded file containing them
2) Take a PKCS#12 encoded file and return a public key object
3) Take a public key object and return a X.509 SubjectKeyInfo encoded binary string

Goal for Friday-12th:
Complete as many requirements under "0.0 - Bring in the public keys!" as possible.

**Comment by srichal** [comment]
Status as of Friday-12th:
Made more progress and completed these steps from story "0.0 - Bring in the public keys!":
4) Take a binary string and return a SHA 256 hash that just returns the top 16 bytes
5) Take a binary string and return a base 64 encoded version
6) Take a PKCS#12 encoded file and return a self signed X.509 cert

Hit two blockers with the below two steps:
7) Show how to submit the PKCS#12 file to create a TLS listening endpoint
8) Show how to submit the PKCS#12 file as a client to create a mutual TLS auth connection

The first blocker was due to the usage of ECDH keys instead of RSA keys. Decided to use RSA for now. The second blocker was due to the self-signed-certificate related "zero-depth-cert" error. I was able to TX the data between server/client but the connection on the client breaks after the data TX. Found the root cause and figured out the solution. Tested the solution using certificates/PCKS12 files created using openssl-command-line. But have yet to implement the solution which needs more code in the C-APi that creates the PKCS12 file.

Goal for Friday-19th:
1) Complete steps 7-to-10 under "0.0 - Bring in the public keys!".
2) Fix my VB-Ubuntu setup so that I can use the add-on functionality of JXcore.
3) Integrate my code into the postcard app.
